### PR TITLE
config: unify behaviors in configuring external libraries

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
This is a reopen of PR #4369 with updates. Fixes issue #4807.

This PR adds a set of `PAC_CHECK_HEADER_LIB_XXX` macros to unify the syntax and behaviors of configuring external libraries.

[skip warnings]

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch3/all
test:mpich/ch4/all

**Comment by hzhou** [comment]
test:mpich/custom --disable-fortran --disable-cxx --disable-romio
netmod: ch4:ofi

**Comment by hzhou** [comment]
test:mpich/custom --disable-fortran --disable-cxx --disable-romio
netmod: ch4:ofi

**Comment by hzhou** [comment]
test:mpich/custom --disable-fortran --disable-cxx --disable-romio
netmod: ch4:ofi

**Comment by hzhou** [comment]
test:mpich/custom --disable-fortran --disable-cxx --disable-romio
netmod: ch4:ofi

**Comment by hzhou** [comment]
test:mpich/ch3/all
test:mpich/ch4/all

**Comment by hzhou** [comment]
test:mpich/custom --disable-fortran --disable-cxx --disable-romio --enable-g=dbg
netmod: ch4:ofi

**Comment by hzhou** [comment]
test:mpich/custom --disable-fortran --disable-cxx --disable-romio --enable-g=dbg
netmod: ch4:ofi

**Comment by hzhou** [comment]
test:mpich/ch3/all
test:mpich/ch4/all

**Review by raffenet — APPROVED** [review]
I wouldn't be surprised if we get tripped up by some corner cases here and there, but this is an overall improvement we can build on.

**Comment by hzhou** [comment]
Thanks @raffenet . I'll do a rebase and update the code for the GPU check, and ask you for another review.

File: configure.ac [hunk 1]
```diff
@@ -1029,21 +1029,15 @@ AC_ARG_VAR([MPLLIBNAME],[can be used to override the name of the MPL library (de
...
-            [with_mpl_prefix=embedded]) dnl action-if-not-given
 mplsrcdir=""
 AC_SUBST([mplsrcdir])
 mpllibdir=""
 AC_SUBST([mpllibdir])
 mpllib=""
 AC_SUBST([mpllib])
-if test "$with_mpl_prefix" = "embedded" ; then
+m4_define([mpl_embedded_dir],[src/mpl])
+PAC_CHECK_HEADER_LIB_EXPLICIT([mpl],[mplconfig.h],[$MPLLIBNAME],[mpl_str_get_int_arg])
```
[the remark this answers is not in the document]
Comment (hzhou) [reply configure.ac:1039]: Sounds good. I can do that.

File: src/pm/hydra/configure.ac [hunk 2]
```diff
@@ -416,33 +389,32 @@ AC_MSG_RESULT([$hydra_topolib_list])
...
-                    AC_CHECK_HEADERS([hwloc.h])
-                    # hwloc_topology_set_pid was added in hwloc-1.0.0, which is our minimum
-                    # required version
-                    AC_CHECK_LIB([hwloc],[hwloc_topology_set_pid])
-                    AC_MSG_CHECKING([if non-embedded hwloc works])
-                    if test "$ac_cv_header_hwloc_h" = "yes" -a "$ac_cv_lib_hwloc_hwloc_topology_set_pid" = "yes" ; then
-                        have_hwloc=yes
-                    fi
-                    AC_MSG_RESULT([$have_hwloc])
+                m4_define([hwloc_embedded_dir],[tools/topo/hwloc/hwloc])
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/pm/hydra/configure.ac:402]: `xxx_embedded_dir` is only used for `m4` logic to check whether the particular library has embedded options. It is used by `m4_ifdef`, thus cannot be a shell variable.

File: src/pm/hydra/configure.ac [hunk 3]
```diff
@@ -142,25 +130,17 @@ AC_SUBST([mpl_srcdir])
...
 AC_SUBST([mpl_lib])
-if test "$with_mpl_prefix" = "embedded" ; then
+m4_define([mpl_embedded_dir],[mpl])
+PAC_CHECK_HEADER_LIB_EXPLICIT([mpl],[mplconfig.h],[$MPLLIBNAME],[mpl_str_get_int_arg])
+if test "$with_mpl" = "embedded" ; then
     mpl_srcdir="mpl"
     mpl_dist_srcdir="mpl"
     mpl_lib="mpl/lib${MPLLIBNAME}.la"
     mpl_subdir_args="--disable-versioning --enable-embedded"
     PAC_CONFIG_SUBDIR_ARGS([mpl],[$mpl_subdir_args],[],[AC_MSG_ERROR(MPL configure failed)])
```
Comment (raffenet) [inline src/pm/hydra/configure.ac:142]: based on this comment: https://github.com/pmodels/mpich/pull/4819/commits/dc6099ec1ef90a4bcdc5ceae0b8221593c8e5337#r568008043

should this be
```
PAC_CONFIG_SUBDIR_ARGS([mpl_embedded_dir],[$mpl_subdir_args],[],[AC_MSG_ERROR(MPL configure failed)])
```

File: src/pm/hydra/configure.ac [hunk 4]
```diff
@@ -142,25 +130,17 @@ AC_SUBST([mpl_srcdir])
 AC_SUBST(mpl_dist_srcdir)
 mpl_includedir=""
 AC_SUBST([mpl_includedir])
-mpl_libdir=""
-AC_SUBST([mpl_libdir])
 mpl_lib=""
 AC_SUBST([mpl_lib])
-if test "$with_mpl_prefix" = "embedded" ; then
+m4_define([mpl_embedded_dir],[mpl])
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/pm/hydra/configure.ac:135]: It is used by `PAC`-macros, there is `m4_ifdef` usage.

File: src/pm/hydra/configure.ac [hunk 5]
```diff
@@ -142,25 +130,17 @@ AC_SUBST([mpl_srcdir])
...
+PAC_CHECK_HEADER_LIB_EXPLICIT([mpl],[mplconfig.h],[$MPLLIBNAME],[mpl_str_get_int_arg])
+if test "$with_mpl" = "embedded" ; then
     mpl_srcdir="mpl"
```
Comment (raffenet) [inline src/pm/hydra/configure.ac:138]: also here. should it be
```
mpl_srcdir="mpl_embedded_dir"
```
like hwloc part does below?

**Comment by hzhou** [comment]
test:mpich/ch3/all
test:mpich/ch4/all
test:mpich/ch4/gpu

File: src/pm/hydra/configure.ac (hunk 3 above)
Comment (hzhou) [reply src/pm/hydra/configure.ac:142]: Hmm, I tried replace `mpl` with `mpl_embedded_dir`, but the result is very difficult to read since we can't tell what happens with the `m4` variable replacement. I'll modify `hwloc` usages below instead, so the m4 variable of `xxx_embedded_dir` is only used for `m4`-logic.

**Comment by hzhou** [comment]
test:mpich/ch3/all
test:mpich/ch4/all
test:mpich/ch4/gpu

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/gpu

**Comment by hzhou** [comment]
test:mpich/ch3/all

**Comment by hzhou** [comment]
test:mpich/ch4/ucx

**Comment by hzhou** [comment]
test:mpich/ch3/all
test:mpich/ch4/most
test:mpich/ch4/gpu

**Comment by hzhou** [comment]
test:mpich/ch3/all
test:mpich/ch4/most
test:mpich/ch4/gpu

**Comment by hzhou** [comment]
test:mpich/ch4/gpu

**Comment by hzhou** [comment]
test:mpich/ch3/all
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch4/gpu

**Comment by hzhou** [comment]
test:mpich/ch4/gpu

**Comment by hzhou** [comment]
test:mpich/ch3/all
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/all
test:mpich/ch4/all

**Comment by hzhou** [comment]
@raffenet I added commits to update the PR for GPU configure. Would you like to review it one more time?

**Review by raffenet — APPROVED** [review]
